src/de.rs

```diff
@@ -556,6 +680,137 @@ impl<S: Read> Decoder<S> {
+        Ok(())
+    }
+
+    async fn ignore_number(&mut self) -> Result<(), Error> {
+        let ch = self.next_char_or_null().await?;
+        match ch {
+            b'0' => {
+                // There cannot be any leading zeroes.
+                // If there is a leading '0', it cannot be followed by more digits.
+                if let b'0'..=b'9' = self.peek_or_null().await? {
+                    return Err(Error::invalid_utf8("invalid number, two leading zeroes"));
+                }
+            }
+            b'1'..=b'9' => {
+                while let b'0'..=b'9' = self.peek_or_null().await? {
+                    self.eat_char().await?;
+                }
+            }
+            _ => {
+                return Err(Error::invalid_utf8(format!(
+                    "invalid number: {}",
+                    self.peek_or_null().await?
+                )));
+            }
+        }
+
+        match self.peek_or_null().await? {
+            b'.' => self.ignore_decimal().await,
+            b'e' | b'E' => self.ignore_exponent().await,
+            _ => Ok(()),
+        }
+    }
+
+    async fn ignore_decimal(&mut self) -> Result<(), Error> {
+        self.next_char().await?;
+
+        let mut at_least_one_digit = false;
+        while let b'0'..=b'9' = self.peek_or_null().await? {
+            self.next_char().await?;
+            at_least_one_digit = true;
+        }
+
+        if !at_least_one_digit {
+            return Err(Error::invalid_utf8(
+                "invalid number, expected at least one digit after decimal",
+            ));
+        }
+
+        match self.peek_or_null().await? {
+            b'e' | b'E' => self.ignore_exponent().await,
+            _ => Ok(()),
+        }
+    }
+
+    async fn ignore_exponent(&mut self) -> Result<(), Error> {
+        self.next_char().await?;
+
+        match self.peek_or_null().await? {
+            b'+' | b'-' => self.eat_char().await?,
+            _ => {}
+        }
+
+        // Make sure a digit follows the exponent place.
+        match self.next_char_or_null().await? {
+            b'0'..=b'9' => {}
+            ch => {
+                return Err(Error::invalid_utf8(format!(
+                    "expected a digit to follow the exponent, found {ch}"
+                )));
+            }
+        }
+
+        while let b'0'..=b'9' = self.peek_or_null().await? {
+            self.eat_char().await?;
+        }
+
+        Ok(())
+    }
+
+    #[async_recursion]
+    async fn ignore_array(&mut self) -> Result<(), Error> {
+        self.eat_char().await?;
+
+        loop {
+            self.expect_whitespace().await?;
+            match self.peek_or_null().await? {
+                b']' => {
+                    self.eat_char().await?;
+                    return Ok(());
+                }
+                b',' => {
+                    self.ignore_exactly(",").await?;
+                }
+                _ => {
+                    self.ignore_value().await?;
+                }
```

Comment:
this should probably be just one arm of the `match`.  i think there would be a bug for
`["test""test"]` would be allowed, when that isn't valid json.

```suggestion
                _ => {
                    self.ignore_value().await?;
                    self.expect_whitespace().await?;
                    self.ignore_exactly(",").await?;
                }
```

Reply:
i think i've fixed this, but haven't put in any tests expecting errors yet